### PR TITLE
Pipeline router tests

### PR DESCRIPTION
**Pull request by dlvenable**
### Description

This PR introduces integration tests for the new router features introduced in Data Prepper 2.0.

This builds on the integration testing support merged in #1949.  Using those tools, this can test pipelines with multiple routes as an integration test using all the Data Prepper core components.
 
### Issues Resolved
N/A
 
### Check List
- [x] New functionality includes testing.
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Comment by dlvenable** [comment]
Now that #1949 is complete, this PR is ready for review.

File: data-prepper-core/src/integrationTest/java/org/opensearch/dataprepper/integration/Router_SingleRouteIT.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    private static final String KNOWN_CONDITIONAL_KEY = "value";
+    private static final String ALPHA_VALUE = "a";
+    private DataPrepperTestRunner dataPrepperTestRunner;
+    private InMemorySourceAccessor inMemorySourceAccessor;
+    private InMemorySinkAccessor inMemorySinkAccessor;
+
+    @BeforeEach
+    void setUp() {
+        dataPrepperTestRunner = DataPrepperTestRunner.builder()
+                .withPipelinesDirectoryOrFile("route/single-route.yaml")
```
Comment (sshivanii) [inline data-prepper-core/src/integrationTest/java/org/opensearch/dataprepper/integration/Router_SingleRouteIT.java:46]: Considering that we'll call a `.yaml` file here is there a reason for calling this method `withPipelinesDirectoryOrFile`?

File: data-prepper-core/src/integrationTest/java/org/opensearch/dataprepper/integration/Router_SingleRouteIT.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+    @Test
+    void sending_non_alpha_events_never_reaches_sink() throws InterruptedException {
+        final List<Record<Event>> randomEvents = createEvents(UUID.randomUUID().toString(), 20);
+
+        Collections.shuffle(randomEvents);
+
+        inMemorySourceAccessor.submit(TESTING_KEY, randomEvents);
+
+        Thread.sleep(1000);
```
Comment (sshivanii) [inline data-prepper-core/src/integrationTest/java/org/opensearch/dataprepper/integration/Router_SingleRouteIT.java:91]: How about we use `wait` and `notifyAll` for thread synchronization instead? Just an alternative to ensure it's not sleeping for too long or not sleeping enough.

File: data-prepper-core/src/integrationTest/java/org/opensearch/dataprepper/integration/Router_ThreeRoutesIT.java [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+        final List<Record<Event>> actualAllRecords = inMemorySinkAccessor.get(ALPHA_SOURCE_KEY);
+
+        assertThat(actualAllRecords.size(), equalTo(alphaEvents.size()));
+
+        assertThat(actualAllRecords, containsInAnyOrder(allEvents.stream()
+                .filter(alphaEvents::contains).toArray()));
+    }
+
+    @Test
+    void sending_alpha_and_beta_events_sends_to_the_sink_with_beta_only_routes() {
```
Comment (chenqi0805) [inline data-prepper-core/src/integrationTest/java/org/opensearch/dataprepper/integration/Router_ThreeRoutesIT.java:118]: This seems redundant with the previous test case to me

File: data-prepper-core/src/integrationTest/java/org/opensearch/dataprepper/integration/Router_SingleRouteIT.java (hunk 2 above)
Comment (dlvenable) [reply data-prepper-core/src/integrationTest/java/org/opensearch/dataprepper/integration/Router_SingleRouteIT.java:91]: The problem with something like that is that we are waiting to ensure that data never reaches the sink. When data reaches the sink I use Awaitility. But, here we just have nothing in particular to wait for. Just waiting long enough for what could process.

File: data-prepper-core/src/integrationTest/java/org/opensearch/dataprepper/integration/Router_ThreeRoutesIT.java (hunk 3 above)
Comment (dlvenable) [reply data-prepper-core/src/integrationTest/java/org/opensearch/dataprepper/integration/Router_ThreeRoutesIT.java:118]: I had wondered about adding this as well. But, it does verify that a second route works in the `routes` list.

```
  - in_memory:
         testing_key: ConditionalRoutingIT_alpha_beta
         routes:
           - alpha
           - beta
```

**Review by chenqi0805 — APPROVED** [review]

File: data-prepper-core/src/integrationTest/java/org/opensearch/dataprepper/integration/Router_ThreeRoutesIT.java (hunk 3 above)
Comment (chenqi0805) [reply data-prepper-core/src/integrationTest/java/org/opensearch/dataprepper/integration/Router_ThreeRoutesIT.java:118]: Wait. I think according to your scenario it should be ALPHA_BETA_SOURCE_KEY in the test body?

**Review by sshivanii — APPROVED** [review]

File: data-prepper-core/src/integrationTest/java/org/opensearch/dataprepper/integration/Router_ThreeRoutesIT.java (hunk 3 above)
Comment (dlvenable) [reply data-prepper-core/src/integrationTest/java/org/opensearch/dataprepper/integration/Router_ThreeRoutesIT.java:118]: Yes, you are correct. These tests are both covering that scenario. I'll create new PR to remove this test and drop that sink entirely.

Comment (dlvenable) [reply data-prepper-core/src/integrationTest/java/org/opensearch/dataprepper/integration/Router_ThreeRoutesIT.java:118]: @chenqi0805, see #2061 for a fix.

**Pull request merged**